sdk/keyvault/Shared/ChallengeBasedAuthenticationPolicy.cs

```diff
@@ -0,0 +1,276 @@
+using System.Threading.Tasks;
+
+namespace Azure.Security.KeyVault
+{
+    internal class ChallengeBasedAuthenticationPolicy : HttpPipelinePolicy
+    {
+        private const string BearerChallengePrefix = "Bearer ";
+        private const string NoChallengeErrorMessage = "Failed to read WWW-Authenticate header from response";
+        private readonly TokenCredential _credential;
+
+        private AuthenticationChallenge _challenge = null;
+
+        private string _headerValue;
+
+        private DateTimeOffset _refreshOn;
+
+        public ChallengeBasedAuthenticationPolicy(TokenCredential credential)
+        {
+            _credential = credential;
+        }
+
+        public override void Process(HttpPipelineMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        {
+            ProcessCoreAsync(message, pipeline, false).GetAwaiter().GetResult();
+        }
+
+        public override Task ProcessAsync(HttpPipelineMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        {
+            return ProcessCoreAsync(message, pipeline, true);
+        }
+
+        private async Task ProcessCoreAsync(HttpPipelineMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
+        {
+            HttpPipelineRequestContent originalContent = message.Request.Content;
+
+            // if this policy doesn't have _challenge cached try to get it from the static challenge cache
+            _challenge ??= AuthenticationChallenge.GetChallenge(message);
+
+            // if we still don't have the challenge for the endpoint
+            // remove the content from the request and send without authentication to get the challenge
+            if (_challenge == null)
+            {
+                message.Request.Content = null;
+            }
+            // otherwise if we already know the challenge authenticate the request
+            else
+            {
+                await AuthenticateRequestAsync(message, async).ConfigureAwait(false);
+            }
+
+            if (async)
+            {
+                await ProcessNextAsync(message, pipeline).ConfigureAwait(false);
+            }
+            else
+            {
+                ProcessNext(message, pipeline);
+            }
+
+            // if we get a 401
+            if (message.Response.Status == 401)
+            {
+                // set the content to the original content in case it was cleared
+                message.Request.Content = originalContent;
+
+                // update the cached challenge
+                var challenge = AuthenticationChallenge.GetChallenge(message);
+
+                if (_challenge == null || challenge != _challenge)
+                {
+                    // update the cached challenge
+                    _challenge = challenge;
+
+                    // if no auth challenge was returned raise an exception since we don't know scopes to authenticate
+                    if (_challenge == null)
```

Comment:
Can we bring this condition out and check it right after `var challenge = AuthenticationChallenge.GetChallenge(message);` or does `f (_challenge == null || challenge != _challenge)` matter here too?